presto-main/src/main/java/com/facebook/presto/metadata/TableHandle.java

```diff
@@ -60,17 +79,30 @@ public boolean equals(Object obj)
         if (this == obj) {
             return true;
         }
-        if (obj == null || getClass() != obj.getClass()) {
+        if (!(obj instanceof TableHandle)) {
             return false;
         }
-        final TableHandle other = (TableHandle) obj;
-        return Objects.equals(this.connectorId, other.connectorId) &&
-                Objects.equals(this.connectorHandle, other.connectorHandle);
+        TableHandle other = (TableHandle) obj;
+        // some ConnectorTableLayoutHandle does not implement equals method, compare isPresent to make iterative optimizer converge
```

Comment:
nit: I edit the comment a bit, feel free to modify it
```
// TODO: This is a big hack.
// Since equals method is not implemented for ConnectorTableLayoutHandle in some connectors, comparing layout might cause iterative optimizer fail to converge. 
// Instead, for now we compare the existence of layout in table handles.
// This works for now since the engine only pushdown filter to connector once during optimization. 
```